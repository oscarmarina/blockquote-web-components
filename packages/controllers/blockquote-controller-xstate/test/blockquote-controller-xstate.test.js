/* eslint-disable import/no-extraneous-dependencies */
import {suite, test, assert, expect, beforeAll, beforeEach} from 'vitest';
import {assert as a11y, fixture, fixtureCleanup} from '@open-wc/testing';
import {getDiffableHTML} from '@open-wc/semantic-dom-diff';
import {html} from 'lit';
import '../demo/xstate-counter.js';

suite('BlockquoteControllerXstate', () => {
  /**
   * * @type {import('../demo/xstate-counter').XstateCounter}
   */
  let el;
  let elShadowRoot;

  suite('Default', () => {
    beforeAll(async () => {
      el = await fixture(html`
        <xstate-counter>light-dom</xstate-counter>
      `);
      elShadowRoot = el?.shadowRoot?.innerHTML;

      return () => {
        fixtureCleanup();
      };
    });

    test('default counter 0', () => {
      assert.equal(el.counterController.snapshot.context.counter, 0);
    });

    test('should get an instance of actor', () => {
      assert.isObject(el.counterController.actor);
    });

    suite('Semantic Dom and a11y', () => {
      test('SHADOW DOM - Structure test', async () => {
        expect(getDiffableHTML(elShadowRoot, {ignoreAttributes: ['id']})).toMatchSnapshot(
          'SHADOW DOM'
        );
      });

      test('LIGHT DOM - Structure test', async () => {
        expect(getDiffableHTML(el, {ignoreAttributes: ['id']})).toMatchSnapshot('LIGHT DOM');
      });

      test('a11y', async () => {
        await a11y.isAccessible(el);
      });
    });
  });

  suite('Events ', () => {
    beforeEach(async () => {
      el = await fixture(html`
        <xstate-counter></xstate-counter>
      `);

      return () => {
        fixtureCleanup();
      };
    });

    test('increases the counter on Increment button click', () => {
      el.shadowRoot?.querySelector('button')?.click();
      assert.equal(el.counterController.snapshot.context.counter, 1);
    });

    test('decreases the counter on Decrement button click', () => {
      el.shadowRoot?.querySelector('button')?.click();
      /** @type {HTMLButtonElement} */ (el.shadowRoot?.querySelector('button + button'))?.click();
      assert.equal(el.counterController.snapshot.context.counter, 0);
    });

    test('disable the counter on Enabled/Disabled button click', () => {
      let lastCounter = el.counterController.snapshot.context.counter;
      const counterButton = el.shadowRoot?.querySelector('button');
      const disabledButton = el.shadowRoot?.querySelector('div + div button');

      counterButton?.click();
      assert.equal(el.counterController.snapshot.context.counter, lastCounter + 1);
      lastCounter = el.counterController.snapshot.context.counter;

      /** @type {HTMLButtonElement} */ (disabledButton)?.click();
      assert.equal(el.counterController.snapshot.value, 'disabled');

      counterButton?.click();
      assert.equal(el.counterController.snapshot.context.counter, lastCounter);
    });
  });
});
