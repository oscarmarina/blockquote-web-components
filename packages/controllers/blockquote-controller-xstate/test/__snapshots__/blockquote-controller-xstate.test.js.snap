// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BlockquoteControllerXstate > Default > Semantic Dom and a11y > LIGHT DOM - Structure test > LIGHT DOM 1`] = `
"<xstate-counter>
  light-dom
</xstate-counter>
"
`;

exports[`BlockquoteControllerXstate > Default > Semantic Dom and a11y > SHADOW DOM - Structure test > SHADOW DOM 1`] = `
"<slot>
</slot>
<div aria-disabled="false">
  <span>
    <button data-counter="increment">
      Increment
    </button>
    <button data-counter="decrement">
      Decrement
    </button>
  </span>
  <p>
    0
  </p>
</div>
<div>
  <button>
    Disabled counter
  </button>
</div>
"
`;
